### PR TITLE
fix: Add mutex fix for delete cname resource

### PR DESCRIPTION
**Pull request by jlilja**
* Adds a mutex lock on the delete for CNAME records.
* Adds tests to verify functionality.
* Adds documentation and examples of data source examples folder.

### CODE REVIEW

File: examples/data-sources/pihole_cname_records/data-source.tf [hunk 1]
```diff
@@ -1 +1,9 @@
+# A data sources to retrieve the existing CNAME records.
 data "pihole_cname_records" "records" {}
+
+# When retrieving a CNAME resource via a data source immediately after creating the resource,
+# It is recommended to set a depends_on constraint in order to let the resource finish
+# creating before performing the fetching of all CNAME records.
+data "pihole_cname_records" "records" {
+    depends_on = [RESOURCE_IDENTIFIER]
+}
```
Comment (ryanwholey) [inline examples/data-sources/pihole_cname_records/data-source.tf:9]: This is a nice addition! My suggestion is to modify it slightly by specifying that this is an issue when retrieving all cname records without any inputs (which it looks like we don't support at the moment). I could see us adding that functionality in the future.

```suggestion
# When retrieving all CNAME records immediately after creating a resource,
# it is recommended to set a depends_on constraint in order to allow the resource finish
# creating before fetching all CNAME records.
resource "pihole_cname_record" "record" {
  domain = "foo.com"
  target = "bar.com"
}

data "pihole_cname_records" "records" {
    depends_on = [pihole_cname_record.record]
}
```

**Review by ryanwholey — COMMENTED** [review]
Thanks for taking a look at this. Would you be interested in opening this as a PR against the upstream?

**Comment by ryanwholey** [comment]
And just to clarify, the theory is that the multiple API requests are made in succession against the API and the API doesn't handle that gracefully?

File: examples/data-sources/pihole_cname_records/data-source.tf (hunk 1 above)
Comment (jlilja) [reply examples/data-sources/pihole_cname_records/data-source.tf:9]: I like it! I added your suggestion.

I agree, supporting that somehow in the future could be useful!

**Comment by jlilja** [comment]
> Thanks for taking a look at this. Would you be interested in opening this as a PR against the upstream?

Absolutely. I messed it up a bit by working solely on the fork and having to wait for current changes in main to be accepted into your repo before mergin this one to my main branch. I suppose that would as well be included in the PR to your repo. Once https://github.com/ryanwholey/terraform-provider-pihole/pull/72 is merged I will create a PR for this one.

After they are done, then I will start working with branches in your repo. :)

> And just to clarify, the theory is that the multiple API requests are made in succession against the API and the API doesn't handle that gracefully?

Correct! I tried to debug this by logging all of the delete requests going to the pihole during the apply, and they reported no errors.

My guess is that when the delete function is called [over here](https://github.com/ryanwholey/terraform-provider-pihole/blob/0abcc9b6c9af74ef80c18ab5eb474bfde73375b5/internal/provider/resource_cname_record.go#L96) then its function code will first try and get all of the current CNAME's via this [line](https://github.com/ryanwholey/terraform-provider-pihole/blob/0abcc9b6c9af74ef80c18ab5eb474bfde73375b5/internal/pihole/cname.go#L136). Maaaybe there could be some race condition that reports that there are no more CNAMEs to be removed, and so it considers its job done.

I'm not sure, and as you said, the _theory_ is that the API can't handle concurrent requests like Terraform does it. I tried to mess limiting its concurrency, which led me down a path of issues that related to mine. Its there I found the issue which I applied to this problem https://github.com/hashicorp/terraform-plugin-sdk/issues/67#issuecomment-1170071965

**Pull request merged**